lib/srv/desktop/rdp/rdpclient/client.go

```diff
@@ -743,6 +743,61 @@ func toClient(handle C.uintptr_t) (value *Client, err error) {
 	return cgo.Handle(handle).Value().(*Client), nil
 }
 
+//export cgo_read_rdp_license
+func cgo_read_rdp_license(handle C.uintptr_t, req *C.CGOLicenseRequest, data_out **C.uint8_t, len_out *C.uint32_t) C.CGOErrCode {
+	client, err := toClient(handle)
+	if err != nil {
+		return C.ErrCodeFailure
+	}
+
+	issuer := C.GoString(req.issuer)
+	company := C.GoString(req.company)
+	productID := C.GoString(req.product_id)
+
+	license, err := client.readRDPLicense(
+		uint16(req.major_version), uint16(req.minor_version),
+		issuer, company, productID)
+	if err != nil {
+		*data_out = nil
+		return C.ErrCodeFailure
+	}
+
+	// in this case, we expect the caller to use cgo_free_rdp_license
+	// when the data is no longer needed
+	*data_out = (*C.uint8_t)(C.CBytes(license))
+	*len_out = C.uint32_t(len(license))
+
+	return C.ErrCodeSuccess
+}
+
+//export cgo_free_rdp_license
+func cgo_free_rdp_license(p unsafe.Pointer) {
+	C.free(p)
+}
+
+//export cgo_write_rdp_license
+func cgo_write_rdp_license(handle C.uintptr_t, req *C.CGOLicenseRequest, data *C.uint8_t, length C.uint32_t) C.CGOErrCode {
+	client, err := toClient(handle)
+	if err != nil {
+		return C.ErrCodeFailure
+	}
+
+	issuer := C.GoString(req.issuer)
+	company := C.GoString(req.company)
+	productID := C.GoString(req.product_id)
+
+	licenseData := asRustBackedSlice(data, int(length))
```

Comment:
Is this to deliberately avoid a copy? We don't really specify how long the backend can keep referencing the value in the item you put in it - and, in fact, the memory backend deliberately doesn't copy anything, so this would cause UB if the `LicenseStore` was pointed at a memorybk instance.

Reply:
Fair point, I'll make a copy so we don't have to worry about it.

Reply:
The `asRustBackedSlice` godoc says "This can be used in lieu of `C.GoBytes` (which copies the data) wherever performance is of greater concern." so I guess you can just change this into `C.GoBytes`.